clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java

```diff
@@ -0,0 +1,166 @@
+import com.netflix.spinnaker.clouddriver.ecs.provider.EcsProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE;
+import static com.netflix.spinnaker.clouddriver.ecs.cache.Keys.Namespace.ECS_CLUSTERS;
+
+abstract class AbstractEcsCachingAgent<T> implements CachingAgent {
+  private final Logger log = LoggerFactory.getLogger(getClass());
+
+  final AmazonClientProvider amazonClientProvider;
+  final AWSCredentialsProvider awsCredentialsProvider;
+  final String region;
+  final String accountName;
+
+  AbstractEcsCachingAgent(String accountName, String region, AmazonClientProvider amazonClientProvider, AWSCredentialsProvider awsCredentialsProvider) {
+    this.accountName = accountName;
+    this.region = region;
+    this.amazonClientProvider = amazonClientProvider;
+    this.awsCredentialsProvider = awsCredentialsProvider;
+  }
+
+  /**
+   * Fetches items from the ECS service.
+   * @param ecs The AmazonECS client that will be used to make the queries.
+   * @param providerCache A ProviderCache that is used to access already existing cache.
+   * @return A list of generic type objects.
+   */
+  protected abstract List<T> getItems(AmazonECS ecs, ProviderCache providerCache);
+
+  /**
+   * Generates a map of CacheData collections associated to a key namespace from a given collection of generic type objects.
+   * @param cacheableItems A collection of generic type objects.
+   * @return A map of CacheData collections belonging to a key namespace.
+   */
+  protected abstract Map<String, Collection<CacheData>> generateFreshData(Collection<T> cacheableItems);
+
+  @Override
+  public String getProviderName() {
+    return EcsProvider.NAME;
+  }
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    String authoritativeKeyName = getAuthoritativeKeyName();
+
+    AmazonECS ecs = amazonClientProvider.getAmazonEcs(accountName, awsCredentialsProvider, region);
+    List<T> items = getItems(ecs, providerCache);
+    return buildCacheResult(authoritativeKeyName, items, providerCache);
+  }
+
+  /**
+   * Provides a set of ECS cluster ARNs.
+   * Either uses the cache, or queries the ECS service.
+   * @param ecs The AmazonECS client to use for querying.
+   * @param providerCache The ProviderCache to retrieve clusters from.
+   * @return A set of ECS cluster ARNs.
+   */
+  Set<String> getClusters(AmazonECS ecs, ProviderCache providerCache) {
+    Set<String> clusters = providerCache.getAll(ECS_CLUSTERS.toString()).stream()
+      .map(cacheData -> (String) cacheData.getAttributes().get("clusterArn"))
+      .collect(Collectors.toSet());
+
+    if (clusters == null || clusters.isEmpty()) {
```

Comment:
`clusters` will never be `null`, will it?

Reply:
Changed - removed `clusters == null`.

You're right - it should never be null. 
`providerCache.getAll(..)` will return an empty set, and the stream will also return an empty set if nothing filters through.